### PR TITLE
Cody: Add support for connecting Cody extension to App

### PR DESCRIPTION
**Pull request by marekweb**
Adds support for connecting the Cody VSCode extension to Sourcegraph App.

### "Connect Sourcegraph App" button

Screenshot 2023-05-11 at 11 53 20 AM

The "Connect Sourcegraph App" feature is behind an experimental flag, `cody.experimental.connectToApp`.

The feature is enabled if App is detected as being installed on the system. This currently works on mac OS. If App is not detected as installed, the button shows "Get Sourcegraph App" instead.

### Local app detector

The `LocalAppDetector` detects whether App is installed. It does this by checking for the existence of known directory locations in the file system.

It uses polling, and for that reason it can be controlled with `start` and `stop` methods. It will stop detecting if the user is authed, or if an App installation has been detected.

### Connection flow

To connect to App, the extension sends the user to App with a sourcegraph:// url to generate a token.

App then sends the user back to the extension using the token callback mechanism.

After successfully connecting:
- the `serverEndpoint` config is set to the local App's endpoint which is `http://localhost:3080`
- the `access-token` secret is set to the access token obtained from App.

## Test plan

- Run a build of Sourcegraph App (it won't work with the dev `sg start app` version)
- Click on "Connect Sourcegraph App" in the extension
- Verify that the flow is working
- Verify that the extension now has a token that works to perform request to the local App instance.

### CODE REVIEW

File: client/cody/webviews/ConnectApp.tsx [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        props.vscodeAPI.postMessage({
+            command: 'links',
+            value: url,
+        })
+    }
+
+    return (
+        <p>
+            {props.isAppInstalled ? (
+                <VSCodeButton
```
Comment (abeatrix) [inline client/cody/webviews/ConnectApp.tsx:22]: small nit 😄  
```suggestion
                <VSCodeButton
                type="button"
                onClick={() =>
                    openLink(
                        props.isAppInstalled
                            ? 'sourcegraph://user/settings/tokens/new/callback?requestFrom=CODY'
                            : 'https://about.sourcegraph.com/app'
                    )
                }
            >
                {props.isAppInstalled ? 'Connect Sourcegraph App' : 'Get Sourcegraph App'}
            </VSCodeButton>
```

**Comment by abeatrix** [comment]
Added cody team as reviewer for second pairs of eyes on the UI 🙇

File: client/cody/src/chat/ChatViewProvider.ts [hunk 2]
```diff
@@ -154,6 +157,9 @@ export class ChatViewProvider implements vscode.WebviewViewProvider, vscode.Disp
                 this.publishConfig()
                 this.sendTranscript()
                 this.sendChatHistory()
+                if (this.config.experimentalConnectToApp) {
```
Comment (abeatrix) [inline client/cody/src/chat/ChatViewProvider.ts:160]: I think we can move this into this.publishConfig(), and then run the `localAppDetector` on startup.

**Review by abeatrix — COMMENTED** [review]
Added some comments and feedback but overall LGTM!
I think we don't need to run the localAppDetector anymore when
- app is detected
- isAuthed = true
wdyt?

File: client/cody/src/chat/ChatViewProvider.ts [hunk 3]
```diff
@@ -610,6 +616,27 @@ export class ChatViewProvider implements vscode.WebviewViewProvider, vscode.Disp
         void this.webview?.postMessage({ type: 'errors', errors: errorMsg })
     }
 
+    public watchLocalAppState(): void {
+        void this.updateLocalAppState()
```
[the remark this answers is not in the document]
Comment (marekweb) [reply client/cody/src/chat/ChatViewProvider.ts:620]: That makes sense, good point!

**Comment by marekweb** [comment]
Thanks @abeatrix -- I rewrote the LocalAppDetector and it now gets started in `publishConfig`.

Now if the extension gets authed, or if experimental.connectToApp is disabled, the detector is stopped; if the detector finds App installed it will also stop itself.

I put the visibility of the "Connect"/"Get" App button behind the experimental.connectToApp flag. So that there's no change to the UI now by default, while we wait until we have a release of App that actually supports the connection.

**Review by abeatrix — APPROVED** [review]
Sounds good and looks good to me 🎉

**Pull request merged**